### DISCUSSION TITLE
Use trained model with torch

### DISCUSSION THREAD

**Discussion by kirillkoncha** [speechbrain/speechbrain#1834] [Q&A]
Hello!

I trained my model for language recognition. Now, I would like to only extract audio embeddings from file with torch (as there is no need for other functions).

However, I am confused as trained model has several checkpoints. I suggest I should use `embedding_model.ckpt`, `normalizer.ckpt`, and `mean_var_norm.ckpt`?

How could I load these models and use them for embedding extraction with torch function? I tried to look into [`class Pretrained`](https://github.com/speechbrain/speechbrain/blob/4b485a32f183e53d7807cbdaff0acc1e93952204/speechbrain/pretrained/interfaces.py#L135), however I did not understand how models are loaded.

I would appreciate if somebody helped me with this.

**Comment by anautsch** [comment]
Hi @kirillkoncha the pretrained interfaces seem apropriate for your need. The Brain class is to train, validate & evaluate models only.

The hyperparameters yaml file will be necessary, so that the pretrained interfaces knows how to use which of the ckpt files you pointed out and what to do with them. For example, here's a LibriSpeech recipe yaml file in different stages:
* for training: https://github.com/speechbrain/speechbrain/blob/develop/recipes/LibriSpeech/ASR/seq2seq/hparams/train_BPE_1000.yaml
* as pretrained model: https://huggingface.co/speechbrain/asr-crdnn-rnnlm-librispeech/blob/main/hyperparams.yaml
* used to train/fine-tune another model: https://github.com/speechbrain/speechbrain/blob/develop/templates/speech_recognition/ASR/train.yaml

There's a copy/paste philosphy between the recipes, which has its history.

This is for the "how to load models" part of your question.

---

After strg+f the pretrained interfaces, how about this one?
https://github.com/speechbrain/speechbrain/blob/4b485a32f183e53d7807cbdaff0acc1e93952204/speechbrain/pretrained/interfaces.py#L859-L861

https://github.com/speechbrain/speechbrain/blob/4b485a32f183e53d7807cbdaff0acc1e93952204/speechbrain/pretrained/interfaces.py#L902-L949

sounds like what you desire to have!

**Reply by kirillkoncha** [reply]
Thank you for your answer! It is pretty much close, however, what I want to do is to see how weights are read and change this process as well (for example, decrypt the weights when reading). Also, what I wanted is to use trained model without the library itself, using only PyTorch

**Reply by anautsch** [reply]
Curious about the decrypt part btw, collaborated on that some time ago.

You're looking into extending the IO process there. Might it be worthwhile to view this as a possible extension to HyperPyYAML and loading pretrained models using it?
cc @pplantinga 

I get you like to use torch directly bc it would be straight to your point. Curious if from there a PR could be derived that helps the SpeechBrain community in return.

**Reply by kirillkoncha** [reply]
Yeah, I also searched HypePyYAML repository trying to understand what code lines are responsible for weights loading, however, did not find anything. In the `class Pretrained` it is difficult to distinguish yaml file reading from model loading.

**Reply by anautsch** [reply]
YAML tells how to read from ckpt. The class needs to be known, so the underlying torch state_dict is interpreted consistently to what's specified. 

YAML -> torch object <- ckpt binary

It comes together here:
https://github.com/speechbrain/speechbrain/blob/4b485a32f183e53d7807cbdaff0acc1e93952204/speechbrain/utils/checkpoints.py#L143-L145

which will be called when, for example, using the  `class Pretrained` (but also when loading from a checkpoint).

**Reply by anautsch** [reply]
Indeed, the HypePyYAML won't help you much, as they are used to instantiate the above `obj` torch.Module. The loading of weights etc happens in the above line 144. That'd be your target :thinking:

**Reply by kirillkoncha** [reply]
Thank you! One more question: how `torch_parameter_transfer` is used in the class? I can't find this function there, but it might be used somewhere in-between other modules and the class.

**Reply by anautsch** [reply]
Tbh - put a debug flag & let it run (came across this today for another issue I had).

https://github.com/speechbrain/speechbrain/blob/4b485a32f183e53d7807cbdaff0acc1e93952204/speechbrain/utils/parameter_transfer.py#L269

https://github.com/speechbrain/speechbrain/blob/4b485a32f183e53d7807cbdaff0acc1e93952204/speechbrain/utils/checkpoints.py#L175-L177

```shell
> grep -rn torch_parameter_transfer speechbrain
speechbrain/utils/checkpoints.py:123:def torch_parameter_transfer(obj, path, device):
speechbrain/utils/checkpoints.py:176:    torch.nn.Module: torch_parameter_transfer,
```
